bob-backend/src/pearl/group.rs

```diff
@@ -335,7 +323,7 @@ impl Group {
     }
 
     pub async fn attach(&self, start_timestamp: u64) -> BackendResult<()> {
-        let holders = self.holders.read().await;
+        let holders = self.holders.upgradable_read().await;
```

Comment:
Same as above: Please add a comment that only one thread can hold the `upgradable_read` lock at a time. This is crucial for proper working

Reply:
Do I understand correctly that there was a bug here before?

Reply:
Yeah, as I understand there was attempt to get write lock inside save_pearl() (inside add()) before read lock was dropped